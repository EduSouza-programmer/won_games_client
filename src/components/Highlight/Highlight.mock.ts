import { HighlightProps } from ".";

export default {
  title: "Read Dead está de volta!",
  subtitle: "Venha conhecer as novas aventuras de John Marston.",
  buttonLabel: "Buy now",
  buttonLink: "https://any",
  backgroundImage: "./img/highlight-back-red-dead.jpg",
  floatImage: "./img/highlight-float-red-dead.png",
} as HighlightProps;

export const highlightExample2 = {
  title: "Borderlands 3!",
  subtitle: "Borderlands está de volta com muitas novidades.",
  buttonLabel: "Buy now",
  buttonLink: "https://any",
  backgroundImage: "./img/highlight-back-borderlands.jpg",
  floatImage: "./img/highlight-float-borderlands.png",
} as HighlightProps;

export const highlightExample3 = {
  title: "Já conhece o clássico CS:GO?",
  subtitle: "Jogue um dos maiores clássicos do FPS.",
  buttonLabel: "Buy now",
  buttonLink: "https://any",
  backgroundImage: "/img/highlight-back-csgo.jpg",
  floatImage: "/img/highlight-float-csgo.png",
} as HighlightProps;
