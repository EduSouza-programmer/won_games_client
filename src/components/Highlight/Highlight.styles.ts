import styled, { css } from "styled-components";
import media from "styled-media-query";

import { HighlightProps } from ".";

type WrapperProps = Pick<HighlightProps, "backgroundImage" | "alignment">;

export const FloatImage = styled.img`
  ${({ theme }) => css`
    z-index: ${theme.zIndex.base};
    max-height: 23rem;
    max-width: 100%;
    grid-area: floatimage;
    align-self: end;
    justify-self: end;

    ${media.greaterThan("medium")`
      max-height: 32rem;
    `}
  `}
`;

export const Content = styled.div`
  ${({ theme }) => css`
    z-index: ${theme.zIndex.base};
    padding: ${theme.spacings.xsmall};
    grid-area: content;

    ${media.greaterThan("medium")`
      align-self: end;
      padding: ${theme.spacings.large};
    `}
  `}
`;

const wrapperModifiers = {
  right: () => css`
    grid-template-areas: "floatimage content";
    grid-template-columns: 1fr 2fr;
    ${Content} {
      text-align: right;
    }
  `,
  left: () => css`
    grid-template-areas: "content floatimage";
    grid-template-columns: 2fr 1fr;
    ${Content} {
      text-align: left;
    }
  `,
};

export const Wrapper = styled.section<WrapperProps>`
  ${({ backgroundImage, alignment }) => css`
    position: relative;
    height: 23rem;
    display: grid;

    background-image: url(${backgroundImage});
    background-position: center center;
    background-size: cover;

    &::after {
      content: "";
      position: absolute;
      width: 100%;
      height: 100%;
      background-color: rgba(0, 0, 0, 0.6);
    }

    ${media.greaterThan("medium")`
      height: 32rem;
    `}

    ${wrapperModifiers[alignment!]()}
  `}
`;

export const Title = styled.h2`
  ${({ theme }) => css`
    font-size: ${theme.typography.fontSizes.large};
    font-weight: ${theme.typography.fontWeightBold};
    color: ${theme.palette.white.main};

    ${media.greaterThan("medium")`
    font-size: ${theme.typography.fontSizes.xxlarge};
    `}
  `}
`;

export const Subtitle = styled.h3`
  ${({ theme }) => css`
    font-size: ${theme.typography.fontSizes.small};
    font-weight: ${theme.typography.fontWeightMedium};
    color: ${theme.palette.white.main};
    margin-bottom: ${theme.spacings.medium};

    ${media.greaterThan("medium")`
    font-size: ${theme.typography.fontSizes.large};
    `}
  `}
`;
