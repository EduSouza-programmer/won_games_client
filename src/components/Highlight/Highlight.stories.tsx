import { Story, Meta } from "@storybook/react/types-6-0";
import { StoryFnReactReturnType } from "@storybook/react/dist/ts3.9/client/preview/types";
import styled from "styled-components";

import highlightMock, {
  highlightExample2,
  highlightExample3,
} from "./Highlight.mock";
import Highlight, { HighlightProps } from ".";

const WrapperCenter = styled.div`
  display: grid;
  align-items: center;
  max-width: 104rem;
  height: 100vh;
  margin: 0 auto;
`;

const defaultDecorator = (Story: Story): StoryFnReactReturnType => (
  <WrapperCenter>
    <Story />
  </WrapperCenter>
);

export default {
  title: "Components/Highlight",
  component: Highlight,
  argTypes: {
    title: {
      type: { required: true },
      description: "Highlight title",
      control: { type: "text" },

      table: {
        category: "texts",
        type: {
          summary: "string",
        },
      },
    },
    subtitle: {
      type: { required: true },
      description: "Hightlight subtitle",
      control: { type: "text" },

      table: {
        category: "texts",
        type: {
          summary: "string",
        },
      },
    },
    buttonLabel: {
      type: { required: true },
      description: "Button label",
      control: { type: "text" },

      table: {
        category: "texts",
        summary: "string",
      },
    },
    buttonLink: {
      type: { required: true },
      description: "Button link",
      control: { type: "text" },

      table: {
        category: "texts",
        summary: "string",
      },
    },
    backgroundImage: {
      type: { required: true },
      description: "Highlight background image",
      control: { type: "text" },

      table: {
        category: "texts",
        summary: "string",
      },
    },
    floatImage: {
      description: "Floating highlight image",
      control: { type: "text" },

      table: {
        category: "texts",
        summary: "string",
      },
    },
    alignment: {
      description: "Image alignment in highlight",
      options: ["left", "right"],
      control: { type: "select" },

      table: {
        category: "choices",
        type: {
          summary: "left | right",
        },
      },
    },
  },
  args: { ...highlightMock },
  parameters: {
    viewport: {
      defaultViewport: "figmaMobile",
    },
    backgrounds: {
      default: "won-dark",
    },
    layout: "fullscreen",
  },
  decorators: [defaultDecorator],
} as Meta<HighlightProps>;

export const Mobile: Story<HighlightProps> = (args) => <Highlight {...args} />;

export const MobileWithoutFloatImage: Story<HighlightProps> = (args) => (
  <Highlight {...args} />
);

MobileWithoutFloatImage.args = {
  floatImage: "",
};

export const Desktop: Story<HighlightProps> = (args) => <Highlight {...args} />;

Desktop.parameters = {
  viewport: {
    defaultViewport: "desktop",
  },
};

export const DesktopWithFloatImage: Story<HighlightProps> = (args) => (
  <Highlight {...args} />
);

DesktopWithFloatImage.parameters = {
  viewport: {
    defaultViewport: "desktop",
  },
};
DesktopWithFloatImage.args = {
  floatImage: "",
};

export const Example2: Story<HighlightProps> = (args) => (
  <Highlight {...args} />
);

Example2.parameters = {
  viewport: {
    defaultViewport: "desktop",
  },
};
Example2.args = {
  ...highlightExample2,
};

export const Example3: Story<HighlightProps> = (args) => (
  <Highlight {...args} />
);

Example3.parameters = {
  viewport: {
    defaultViewport: "desktop",
  },
};
Example3.args = {
  ...highlightExample3,
};
