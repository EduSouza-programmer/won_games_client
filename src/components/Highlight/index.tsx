import Image from "next/image";

import Heading from "@/components/Heading";
import Button from "@/components/Button";
import * as S from "./Highlight.styles";

export type HighlightProps = {
  title: string;
  subtitle: string;
  buttonLabel: string;
  buttonLink: string;
  backgroundImage: string;
  floatImage?: string;
  alignment?: "left" | "right";
};

function Highlight({
  title,
  subtitle,
  buttonLabel,
  buttonLink,
  backgroundImage,
  floatImage,
  alignment = "right",
}: HighlightProps) {
  return (
    <S.Wrapper alignment={alignment}>
      <S.BackgroundImage
        layout="fill"
        src={backgroundImage}
        placeholder="blur"
        blurDataURL={backgroundImage}
        alt={`A background image with title ${title}`}
        objectFit="cover"
      />

      <S.Grid className="oi">
        {!!floatImage && (
          <S.FloatImage>
            <Image width={290} height={270} src={floatImage} alt={title} />
          </S.FloatImage>
        )}
        <S.Content>
          <Heading
            responsiveSize={{
              min: 16,
              max: 30,
              isNoWrapperText: true,
            }}
          >
            {title}
          </Heading>
          <S.Subtitle
            as="h3"
            className="subtitle"
            responsiveSize={{ min: 11, max: 20 }}
          >
            {subtitle}
          </S.Subtitle>
          <Button as="a" href={buttonLink}>
            {buttonLabel}
          </Button>
        </S.Content>
      </S.Grid>
    </S.Wrapper>
  );
}

export default Highlight;
