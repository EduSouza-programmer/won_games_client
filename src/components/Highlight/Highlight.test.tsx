import { screen } from "@testing-library/react";
import { renderWithTheme } from "@/utils/tests/helpers";

import * as S from "./Highlight.styles";

import Highlight from ".";

const propsRequired = {
  title: "Heading 1",
  subtitle: "Heading 2",
  buttonLabel: "Buy now",
  buttonLink: "https://any",
  backgroundImage: "/img/red-dead-img.jpg",
};

describe("<Highlight />", () => {
  it("should render the headings and button", () => {
    renderWithTheme(<Highlight {...propsRequired} />);

    expect(
      screen.getByRole("heading", { name: /heading 1/i })
    ).toBeInTheDocument();
    expect(
      screen.getByRole("heading", { name: /heading 2/i })
    ).toBeInTheDocument();
    expect(screen.getByRole("link", { name: /Buy now/i })).toHaveAttribute(
      "href",
      "https://any"
    );
  });

  it("should render background image", () => {
    const { container } = renderWithTheme(<Highlight {...propsRequired} />);

    expect(container.firstChild).toHaveStyle({
      backgroundImage: `url(${propsRequired.backgroundImage})`,
    });
  });

  it("should render float image by props passed", () => {
    renderWithTheme(
      <Highlight {...propsRequired} floatImage="/img/red-dead-float.png" />
    );

    expect(
      screen.getByRole("img", { name: propsRequired.title })
    ).toHaveAttribute("src", "/img/red-dead-float.png");
  });

  it("should render align right S.Content correcty by default", () => {
    const { container } = renderWithTheme(<Highlight {...propsRequired} />);

    expect(container.firstChild).toHaveStyleRule(
      "grid-template-areas",
      /floatimage content/
    );

    expect(container.firstChild).toHaveStyleRule("text-align", "right", {
      modifier: `${S.Content}`,
    });
  });

  it("should render align left S.Content correcty when props passed", () => {
    const { container } = renderWithTheme(
      <Highlight {...propsRequired} alignment="left" />
    );

    expect(container.firstChild).toHaveStyleRule(
      "grid-template-areas",
      /content floatimage/
    );

    expect(container.firstChild).toHaveStyleRule("text-align", "left", {
      modifier: `${S.Content}`,
    });
  });
});
