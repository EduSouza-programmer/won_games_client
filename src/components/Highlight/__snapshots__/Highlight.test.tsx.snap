// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Highlight /> should render the headings and button 1`] = `
.c5 {
  color: #FAFAFA;
  font-size: calc( 16px + (30 - 16) * ((100vw - 320px) / (1920 - 320)) );
  white-space: nowrap;
}

.c6 {
  color: #FAFAFA;
  font-size: calc( 11px + (20 - 11) * ((100vw - 320px) / (1920 - 320)) );
}

.c8 {
  background: linear-gradient(180deg,#ff5f5f 0%,#f062c0 50%);
  color: #FAFAFA;
  border: 0;
  border-radius: 0.4rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  height: 4rem;
  font-size: 1.4rem;
  padding: 0.8rem 3.2rem;
}

.c8:hover {
  background: linear-gradient(180deg,#e35565 0%,#d958a6 50%);
}

.c4 {
  z-index: 10;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  grid-area: content;
}

.c4 .subtitle {
  font-weight: 400;
  margin-bottom: 1.6rem;
}

.c4 .c7 {
  box-shadow: 0px 6px 6px -3px rgba(0,0,0,0.2),0px 10px 14px 1px rgba(0,0,0,0.14),0px 4px 18px 3px rgba(0,0,0,0.12);
}

.c2 {
  display: grid;
  height: 27rem;
}

.c0 {
  position: relative;
}

.c0::after {
  content: "";
  position: absolute;
  inset: 0;
  background-color: rgba(0,0,0,0.6);
}

.c0 .c1 {
  grid-template-areas: "floatimage content";
  grid-template-columns: 1fr 1.5fr;
}

.c0 .c1 .sc-dlnjwi {
  justify-self: flex-start;
}

.c0 .c1 .c3 {
  justify-self: end;
  padding-top: 1.6rem;
  padding-bottom: 1.6rem;
  padding-right: 1.6rem;
}

.c0 .c1 .c3 .c7 {
  -webkit-align-self: flex-end;
  -ms-flex-item-align: end;
  align-self: flex-end;
}

@media (min-width:1920px) {
  .c5 {
    font-size: 30px;
  }
}

@media (min-width:1920px) {
  .c6 {
    font-size: 20px;
  }
}

@media (min-width:769px) {
  .c4 {
    -webkit-align-self: end;
    -ms-flex-item-align: end;
    align-self: end;
  }
}

@media (min-width:450px) {
  .c4 .subtitle {
    margin-bottom: 3.2rem;
  }
}

@media (min-width:769px) {
  .c2 {
    height: 32rem;
  }
}

@media (max-width:450px) {
  .c0 .c1 .sc-dlnjwi > div {
    left: 0;
    bottom: 0;
  }
}

@media (min-width:769px) {
  .c0 .c1 .c3 {
    padding-top: 4.8rem;
    padding-bottom: 4.8rem;
    padding-right: 4.8rem;
  }
}

<div
  class="c0"
>
  <img
    alt="A background image with title Heading 1"
    layout="fill"
    placeholder="blur"
    src="/img/red-dead-img.jpg"
  />
  <div
    class="c1 c2 oi"
  >
    <div
      class="c3 c4"
    >
      <h2
        class="c5"
        color="white"
      >
        Heading 1
      </h2>
      <h3
        class="c6 subtitle"
        color="white"
      >
        Heading 2
      </h3>
      <a
        class="c7 c8"
        href="https://any"
      >
        <span>
          Buy now
        </span>
      </a>
    </div>
  </div>
</div>
`;
